### ISSUE TITLE
App list is not complete

### ISSUE THREAD

**Issue by MPeti1** [MuntashirAkon/AppManager#140]
**Describe the bug**
On the main menu (where the installed apps are listed) there are certain apps that I can't find.
One such app is the Reddit client named Boost (com.rubenmayayo.reddit).

**To Reproduce**
Steps to reproduce the behaviour:

0. Install the Boost app
1. Open App Manager
2. Scroll to the apps that start with the letter B, or search for Boost
3. The mentioned app is not there

**Expected behavior**
The app should be listed there

**Screenshots**
Please click on the images to enlarge them

Settings app info page of Boost | Boost not found by App Manager | Boost not found by Skit | Boost found by Titanium Backup | Boost's about page
:-------------------------------------:|:----------------------------------------:|:---------------------------:|:-----------------------------------------:|:-----------------------:
![photo_2020-11-01_16-34-46](https://user-images.githubusercontent.com/9215491/97807377-a83de980-1c60-11eb-9124-e0d15982d69e.jpg) | ![photo_2020-11-01_16-35-13](https://user-images.githubusercontent.com/9215491/97807431-fbb03780-1c60-11eb-9d3a-4c7dea4f6232.jpg) | ![photo_2020-11-01_16-35-01](https://user-images.githubusercontent.com/9215491/97807387-b7bd3280-1c60-11eb-9b3a-1d582d7e745e.jpg) | ![photo_2020-11-01_16-35-17](https://user-images.githubusercontent.com/9215491/97807492-82651480-1c61-11eb-8b09-19bc4c40d3ac.jpg) | ![photo_2020-11-01_16-35-21](https://user-images.githubusercontent.com/9215491/97807520-9b6dc580-1c61-11eb-80f3-d92a8d26fad1.jpg)

**Crash logs**
There weren't any, but previously I had, 2 versions earlier. When I tried changing the theme to dark or battery mode, after I reopened the app it was crashing. It had found a limited list of apps, and it crashed with an exception signaling that the Pacakge Manager service died, which I haven't seen earlier.
I'm a new user, and I came on that version, so whatever was it, it might haven't been introduced on that version.

**Device info**
 - Device: Xiaomi Redmi 3S
 - OS Version: MIUI 8
 - App Manager Version: 25.5.17 (368)
 - Mode: no-root

**Additional context**
As you can see in the screenshots, there are other app managers too that won't list Boost, but it varies. Like, there are one named MyAPK (com.andatsoft.myapk.fwa), which sees it, but Skit doesn't

Also, keep up the good work! I like your app manager very much, and the new Scanner feature is very good, I like it!
Did you think about making that part to be a library? Earlier I was thinking about making an app, that either works as an Xposed module, or by modifying the apks, that would disable libraries (mostly tracking ones) or straight out remove them, or at least part of them, but I haven't gotten to making it, mostly because I haven't known of a way to identify libraries

**Comment by MuntashirAkon** [comment]
Unfortunately, I cannot reproduce this issue. If you're interested, you can do something like this:
1. Uninstall Boost app
2. Open App Manager (and keep it open in background) and sort apps by last update
3. Install Boost app again
4. Check if Boost app is present in the first row
5. If not present, close App Manager (don't open it again!)
6. Send me the file located at `/sdcard/Android/data/io.github.muntashirakon.AppManager/files/cache/am.log` (you can send it to me via PM on Telegram or via email if you're concerned about privacy)

> There weren't any, but previously I had, 2 versions earlier. When I tried changing the theme to dark or battery mode, after I reopened the app it was crashing. It had found a limited list of apps, and it crashed with an exception signaling that the Pacakge Manager service died, which I haven't seen earlier.

Yes, the issue has been fixed in the latest version. Android allows a limited number of calls to PackageManager in a given time.

> Did you think about making that part to be a library? Earlier I was thinking about making an app, that either works as an Xposed module, or by modifying the apks, that would disable libraries (mostly tracking ones) or straight out remove them, or at least part of them, but I haven't gotten to making it, mostly because I haven't known of a way to identify libraries

My library scanning isn't perfect though. There are still a lot of things to do (see #126). There are also other problems that I think is beyond this project such as bypassing obfuscation (like LibRadar or LibScout). I also have other plans for this page such as #133.

**Comment by MPeti1** [comment]
> Install Boost app again

Is it ok if I do the reinstalling in the form of restoring a backup (with Titanium Backup)?

> My library scanning isn't perfect though. There are still a lot of things to do (see #126). There are also other problems that I think is beyond this project such as bypassing obfuscation (like LibRadar or LibScout). I also have other plans for this page such as #133.

Hmm you have some interesting plans here! :)
I understand that bypassing obfuscation might not be of priority now, but I think [ClassHunter](https://github.com/Nordskog/ClassHunter/) could help you in solving that. I haven't heard of the other 2, I'll check them out, thank you!

**Comment by MPeti1** [comment]
Restoring haven't fixed it, but I'm unable to install the app through the play store now. I was in the beta test program there, and the store first wants me to remove from the program for some reason, and it can take hours in my experience

**Comment by MuntashirAkon** [comment]
> it's installer is `Market Feedback Agent`, is it ok?

Installer is provided by the system. It is like that probably because TB set it during the restore process.

**Comment by MPeti1** [comment]
> Installer is provided by the system. It is like that probably because TB set it during the restore process.

Oh, thought that it's always the name of the installer app, good to know it can be changed by the installer.

I just noticed that the app has gone missing from the list again. I now uninstalled and then installed it, and for a moment it was shown on the main page, but after refreshing it disappeared again.

After following points 5 and 6, I've sent you the logs on Telegram, though the file does not seem to contain events related to app installation or removal.
Also, on installation Boost always appears in the list automatically (probably you listen for app install events), but only disappears if I refresh or reload the app, so it seems that the system is not sending out relevant events, and it just won't include it in package queries anymore

Also, a note about the log file: since opening the issue, I've flipped on root mode, allowed by Magisk.  
The network error is just that I disabled network access after trying out ADB mode. I know with root it doesn't matter much, but I like to keep it that way anyways for apps that don't need network connection.

**Comment by MuntashirAkon** [comment]
> The network error is just that I disabled network access after trying out ADB mode. I know with root it doesn't matter much, but I like to keep it that way anyways for apps that don't need network connection.

Sorry, the documentation is a bit outdated about this. You can disable network access in ADB mode as well. AM uses socket connection to localhost (127.0.01) which isn't affected by AFWall+ or Netguard.

**Comment by MPeti1** [comment]
No worries! Is it a setting, or is it just not necessary?
Also, I only mentioned it so you know that it's not a bug, just the reaction of the system.

**Comment by MPeti1** [comment]
In the meantime I think I've found more such apps: The [Gallery](https://f-droid.org/en/packages/com.simplemobiletools.gallery.pro/) and the [Contacts](https://f-droid.org/en/packages/com.simplemobiletools.contacts.pro/) app from SimpleMobileTools is also missing

SMT apps listed as installed in F-droid | Only Simple Calendar is shown in App Manager
:-:|:-:
![kép](https://user-images.githubusercontent.com/9215491/97872887-0e854380-1d17-11eb-8dd9-da9670df918e.png)|![kép](https://user-images.githubusercontent.com/9215491/97873099-6b80f980-1d17-11eb-8d36-85f3d2960af8.png)

**Comment by MuntashirAkon** [comment]
Hmm, there must be some issues with the PackageManager. Check if these packages are present in **app_Packages Info** (available on F-Droid). If not, check if they can be accessible via the `pm` command:
```
pm list packages
```

> No worries! Is it a setting, or is it just not necessary?
> Also, I only mentioned it so you know that it's not a bug, just the reaction of the system.

I'm not sure actually. AM uses `app_process` to access the hidden APIs accessible only to the system apps. This is a sort of hack-y workaround but it still works. Using `app_process` command, AM runs a daemon in an environment where privileged instructions can be executed. To communicate with the daemon, it need a socket connection. In order to establish a socket connection `android.permission.INTERNET` is required.

**Comment by MPeti1** [comment]
app_Packages Info can see the apps

Boost | SImple apps
:-:|:-:
![image](https://user-images.githubusercontent.com/9215491/98036892-10d0c600-1e1b-11eb-91e1-2819a3263fae.png) | ![image](https://user-images.githubusercontent.com/9215491/98036548-82f4db00-1e1a-11eb-9d9a-10afe009b353.png)

**Comment by MPeti1** [comment]
pm lists the packages too:

![image](https://user-images.githubusercontent.com/9215491/98037244-90f72b80-1e1b-11eb-87ba-5910d984c439.png)

**Comment by MuntashirAkon** [comment]
So, my guess is this is a problem with my algorithm or a problem with PackageManager. To confirm it, download one of the latest debug builds from [Github actions](https://github.com/MuntashirAkon/AppManager/actions). It has a different package name, so you can run side by side with the regular AM. In the main page, click on the three-dots menu and select profiles. Using the plus button create a new profile (and a new page will open). Now, click on the plus button. After some time, a dialog will open. Use the search button in the dialog to see if these apps are present there.

**Comment by MPeti1** [comment]
The apps are present in this menu with [this](https://github.com/MuntashirAkon/AppManager/actions/runs/357967840) build

Boost in the list | Simple apps in the list | All of these selected
:-:|:-:|:-:
![image](https://user-images.githubusercontent.com/9215491/98876036-aef90780-247d-11eb-9af6-fa065bf0bc3d.png) | ![image](https://user-images.githubusercontent.com/9215491/98876045-b4565200-247d-11eb-89f9-1202909b908c.png) | ![image](https://user-images.githubusercontent.com/9215491/98876052-ba4c3300-247d-11eb-8f30-5b2c3ec51cbb.png)

**Comment by MuntashirAkon** [comment]
Install the apk below and open it. After that send me the am.log file located at 
```
/sdcard/Android/data/io.github.muntashirakon.AppManager.debug/files/cache/am.log
```
[app-universal-debug.apk.zip](https://github.com/MuntashirAkon/AppManager/files/5528610/app-universal-debug.apk.zip)

**Comment by MPeti1** [comment]
Here is the file after installing thank and running the debug app: [am.log](https://github.com/MuntashirAkon/AppManager/files/5533844/am.log)

**Comment by MuntashirAkon** [comment]
> Here is the file after installing thank and running the debug app: [am.log](https://github.com/MuntashirAkon/AppManager/files/5533844/am.log)

Weird. Logs are saying that all 254 packages are loaded.

**Comment by MPeti1** [comment]
Could I help with live debugging the app? Where would you put breakpoints to see what partial results are?  
I'm a bit tight on time nowadays, so I can't dive into the code, but if you tell me what should I look for and where, I think I could help

**Comment by MuntashirAkon** [comment]
> Could I help with live debugging the app? Where would you put breakpoints to see what partial results are?

It won't be necessary for now. I'll send you another debug build later.

**Comment by MuntashirAkon** [comment]
@MPeti1 Install the previously sent debug build again. This time, long click on any app and then use the select all button to select all apps to get a count. This count shouldn't match with the ones displayed in am.log.

**Comment by MPeti1** [comment]
I've installed [this](https://github.com/MuntashirAkon/AppManager/issues/140#issuecomment-725864983) again, but both the log and the count in the menu shows 253. I've deleted an app since the last comment, so the one less number is not a problem

While it's not probable that your app is affected by it, because it's fairly new and I haven't installed new modules in a while, you should know that I have the Xposed Framework installed.  
It makes it possible that an app modifies the behavior of an other one, or even of the system, but it can only be used for Java code.

Long story short, in some cases it can be detected by the current stack trace if the currently executing method is hooked.
If you suspect that things happen that in no way should, you could try logging the stack trace there, and I'll see if the logs contain anything suspicious

**Comment by MuntashirAkon** [comment]
> but both the log and the count in the menu shows 253.

Are those apps still missing?

**Comment by MPeti1** [comment]
In the debug app, yes

**Comment by MuntashirAkon** [comment]
> Long story short, in some cases it can be detected by the current stack trace if the currently executing method is hooked.
> If you suspect that things happen that in no way should, you could try logging the stack trace there, and I'll see if the logs contain anything suspicious

Since both counts are 253, there aren't any apps missing. So, no stack trace.

**Comment by MPeti1** [comment]
The apps are still missing. Both `pm list | wc -l` and ShowJava's loading screen (when it's loading the list of apps to pick from) shows 464. Titanium Backup shows 500+, but that's probably because it counts in uninstalled but backed up apps, too

**Comment by MuntashirAkon** [comment]
> The apps are still missing. Both `pm list | wc -l` and ShowJava's loading screen (when it's loading the list of apps to pick from) shows 464.

You probably mean `pm list packages | wc -l`. Still, my guess is that this is happening due a problem with the package manager (could be due to patched `core.jar` or Xposed). Follow the instruction of [this comment](https://github.com/MuntashirAkon/AppManager/issues/140#issuecomment-725485980) but for the attached apk. If my guess is correct, you wouldn't find these apps in the profile section this time.

[app-universal-debug.apk.zip](https://github.com/MuntashirAkon/AppManager/files/5578675/app-universal-debug.apk.zip)

**Comment by MPeti1** [comment]
Sorry for the delay. Yes, I don't find the apps there now.  
So it seems the problem originates from the Package Manager, and I'm sorry for taking your time with a system quirk

**Comment by MuntashirAkon** [comment]
> Sorry for the delay. Yes, I don't find the apps there now.

Thanks for the reply.

> So it seems the problem originates from the Package Manager

Yes, indeed. This can be fixed but the cost is too much. I'll see if I can find a better solution.

**Comment by MuntashirAkon** [comment]
So far it has been confirmed that one or more of the three flags (`GET_SIGNATURES`, `GET_ACTIVITIES` and `MATCH_DISABLED_COMPONENTS`) are causing the issues. However, further study is necessary to actually figure out which of them are causing issues.

But due to several other issues, AM will use its own package parser instead of relying on Android's package parser in future which will solve this issue and partly solve #101, #167, merge of autostarts and activity shortcuts.

**Comment by MPeti1** [comment]
Thank you! I didn't expect you to go out of your way to fix bugs of older systems, I mean it's reasonable that if someone develops an app, if something doesn't work as documented for an individual user than it's hard to "trust" that the system will work as expected in any other ways, and would be a waste of resources to try fixing it, but it's interesting that there are other users too with a similar problem

**Comment by MuntashirAkon** [comment]
> I didn't expect you to go out of your way to fix bugs of older systems

Android 5 is still a widely used operating system, mainly used as a simulator. I know at least ten people who're still using this OS as their daily driver. So, supporting this operating system is a necessity for apps like AM whose main focus is the people involved or interested in IT.

**Comment by MPeti1** [comment]
Thank you for this attitude! It's good to know there are still devs like you.


<details>
  <summary>A bit off topic, but an explanation about what I mean by this:</summary>
Recently I asked the devs of Termux why did they drop support for Android versions below 7. Turns out it's because their app crashes on those versions, because google play tampers with an .so file of theirs (for an unknown reason).

I don't get why does a native android terminal emulator need bootstrap, but that's not the only problem. The idea that google play tampers with an .so file in unknown ways and for unknown reasons is horrible in itself, and even more, the devs doesn't seem to be  interested in finding out what happens actually.
</details>

**Comment by MuntashirAkon** [comment]
Termux is a very sophisticated app, and compatibility on Android is a nightmare. I can get past many compatibility issues by handling them inside AM instead of the OS itself but Termux cannot do that. The latest and, perhaps, the greatest one is the enforcement of W^X on Android (ie. you cannot run any executables from an apps data folder). Another issue is that they use Google Play Store which is only good at imposing weird rules and distributing malware.

**Comment by MPeti1** [comment]
I have good news. I've just installed 2.5.19, and now I see the apps that were missing previously!
This wasn't the case in 2.5.18, so the change in effect must have been made in .19. Did you make some related changes? I don't see similar ones in the changelog

**Comment by MuntashirAkon** [comment]
> Did you make some related changes?

No.

**Comment by MuntashirAkon** [comment]
Anyway, this could be due to some unrelated changes in MainViewModel. Try one of the debug builds to make sure that it's still the case.

**Comment by MPeti1** [comment]
It's still ok in Build #557

**Comment by MuntashirAkon** [comment]
We can safely close this issue, I think.
